src/EditorFeatures/Core.Wpf/Options/EditorConfigDocumentOptionsProvider.cs

```diff
@@ -63,6 +65,69 @@ private void Workspace_DocumentOpened(object sender, DocumentEventArgs e)
             }
         }
 
+        private void Workspace_WorkspaceChanged(object sender, WorkspaceChangeEventArgs e)
+        {
+            switch (e.Kind)
+            {
+            case WorkspaceChangeKind.SolutionRemoved:
+            case WorkspaceChangeKind.SolutionCleared:
+                ClearOpenFileCache();
+                break;
+
+            case WorkspaceChangeKind.ProjectRemoved:
+                ClearOpenFileCache(e.ProjectId);
+                break;
+
+            default:
+                break;
+            }
+        }
+
+        private void ClearOpenFileCache(ProjectId projectId = null)
+        {
+            lock (_gate)
+            {
+                var itemReleased = false;
+                foreach (var (documentId, contextTask) in _openDocumentContexts)
+                {
+                    if (projectId is null || documentId.ProjectId == projectId)
+                    {
+                        itemReleased = true;
+                        ReleaseContext_NoLock(contextTask);
+                    }
+                }
+
+                // If any items were released due to the Clear operation, we need to remove them from the map.
+                if (itemReleased)
+                {
+                    if (projectId is null)
+                    {
+                        _openDocumentContexts.Clear();
+                    }
+                    else
+                    {
+                        var itemsToRemove = _openDocumentContexts.Keys.Where(key => key.ProjectId == projectId).ToList();
```

Comment:
If we're still creating a list here, why not just add the keys to release in a list in the first loop and avoid having the duplicate logic? I get we can avoid the list in the close solution case, but the number of open files isn't usually going to be big enough for that to matter.

Reply:
➡️ Replaced with simpler logic in 0473ba4f0089e72a052147f7efeeb6818c03c5c5.